WT-15495 Expose custom compilation flags to user

ablekhman opened the pull request:
- Extra compilation flags now added to the standard `CMAKE_<LANG>_FLAGS_<CONFIG>` variables instead of being silently added to the directory
- Flags are visible to the user during config phase
- User can edit the flags, if needed

ablekhman committed:
WT-15495 Expose extra debug compilation flags to user

- Extra compilation flags now added to the standard
    `CMAKE_<LANG>_FLAGS_<CONFIG>` variables instead
    of being silently applied to directories
- Flags are visible to the user during config phase
- User can edit the flags, if needed

ablekhman commented on cmake/configs/x86/windows/config.cmake:
```diff
@@ -8,8 +8,6 @@ set(ENABLE_STATIC ON CACHE BOOL "" FORCE)
 set(ENABLE_SHARED OFF CACHE BOOL "" FORCE)
 set(WITH_PIC ON CACHE BOOL "" FORCE)
 
-# Inline expansion.
```
This is not needed. Default optimisation levels control inline expansion already.

ablekhman commented on cmake/configs/base.cmake:
```diff
@@ -367,8 +367,6 @@ config_string(
     DEFAULT "${default_optimize_level}"
 )
 
-add_compile_options("${CC_OPTIMIZE_LEVEL}")
```
The flags used later with final `CMAKE_*` variables in modes.cmake.

ablekhman commented on cmake/configs/base.cmake:
```diff
@@ -400,24 +398,44 @@ endif()
 if(ENABLE_DEBUG_INFO)
-    if("${CMAKE_C_COMPILER_ID}" STREQUAL "MSVC")
-        # Produce full symbolic debugging information.
-        add_compile_options(/Z7)
```
Redundant these days. MSVC produces full debug information with default flags already.

ablekhman commented on cmake/configs/base.cmake:
```diff
@@ -436,12 +454,14 @@ endif()
+    # Define runtime library flags based on DYNAMIC_CRT setting
     if(DYNAMIC_CRT)
         # Use the multithread-specific and DLL-specific version of the run-time library (MSVCRT.lib).
-        add_compile_options(/MD)
```
CRT usage flags set properly this time. It also fixes numerous warnings (due to flag conflicts) in Windows builds.

lukech commented on cmake/configs/base.cmake:
```diff
@@ -400,24 +398,44 @@ endif()
     endif()
+
+    # MSVC: ensure linker produces PDBs.
+    if("${CMAKE_C_COMPILER_ID}" STREQUAL "MSVC")
```
Can this line be replaced with `if(MSVC_CXX_COMPILER)`, similar as what's been done above in CMakeLists.txt?

lukech commented on cmake/configs/modes.cmake:
```diff
@@ -205,23 +205,24 @@ endif()
 if("${CMAKE_C_COMPILER_ID}" STREQUAL "MSVC")
-    string(REPLACE "/O3" "" CMAKE_C_FLAGS_RELEASE ${CMAKE_C_FLAGS_RELEASE})
-    string(REPLACE "/O3" "" CMAKE_CXX_FLAGS_RELEASE ${CMAKE_CXX_FLAGS_RELEASE})
-    string(REPLACE "/Z7" "" CMAKE_C_FLAGS_RELWITHDEBINFO ${CMAKE_C_FLAGS_RELWITHDEBINFO})
-    string(REPLACE "/Z7" "" CMAKE_CXX_FLAGS_RELWITHDEBINFO ${CMAKE_CXX_FLAGS_RELWITHDEBINFO})
```
Should `/Z7` be set as `debug_flags`?

ablekhman replied:
MSVC sets `/Zi` by default for all relevant configurations. We don't need to worry about the flag explicitly.
We could replace default `/Zi` (.pdb file generated separately) with `/Z7` (.pdb info is embedded), but I don't see the advantage of doing it. That's why I left default settings intact.

lukech replied:
Got it, thanks

lukech commented on cmake/helpers.cmake:
```diff
@@ -589,23 +589,88 @@ function(parse_filelist_source filelist output_var)
-    escape_regex_special_characters("${flag}" escaped_flag)
-    if (NOT ${included_flags} MATCHES ".*${escaped_flag}.*")
+    string(FIND "${${included_flags}}" "${flag}" flag_position)
+    if(flag_position EQUAL -1)
```
Love this simplification!

ablekhman committed:
WT-15495 Expose extra debug compilation flags to user

- Extra compilation flags now added to the standard
    `CMAKE_<LANG>_FLAGS_<CONFIG>` variables instead
    of being silently applied to directories
- Flags are visible to the user during config phase
- User can edit the flags, if needed

ablekhman committed:
WT-15495 Expose extra debug compilation flags to user

- Extra compilation flags now added to the standard
    `CMAKE_<LANG>_FLAGS_<CONFIG>` variables instead
    of being silently applied to directories
- Flags are visible to the user during config phase
- User can edit the flags, if needed

lukech reviewed (approved):
LGTM

ivankochin commented on cmake/configs/base.cmake:
```diff
@@ -82,6 +82,7 @@ config_bool(
     ENABLE_ANTITHESIS
     "Enable the Antithesis random library"
     DEFAULT OFF
+    DEPENDS "WT_POSIX"
```
Is Antithesis supported on both Linux and MacOS?

ablekhman replied:
I picked this condition from here: https://github.com/wiredtiger/wiredtiger/blob/develop/cmake/install/install.cmake#L41
Just wanted to make it consistent.

ivankochin commented on cmake/configs/base.cmake:
```diff
@@ -400,24 +399,44 @@ endif()
+        add_cmake_compiler_flags(
+            FLAGS -g3 -gdwarf-4
+            LANGUAGES C CXX
+            BUILD_TYPES Debug RelWithDebInfo
```
Should we define these flags for sanitizer build types (TSan, MSan, ...)? Is there a way to define compiler flags for all build types (maybe except release)?

WDYT about introducing `add_cmake_debug_info_compiler_flags` that would effectively use `add_cmake_compiler_flags` specifying the required LANGUAGES and BUILD_TYPES?

ablekhman replied:
Good point about other build types. I changed the code to include sanitiser builds.

ivankochin commented on cmake/configs/base.cmake:
```diff
@@ -400,24 +399,44 @@ endif()
+            endif()
+            set(${link_var} "${${link_var}}" CACHE STRING "Linker flags for executables" FORCE)
+        endforeach()
+    endif()
```
Would something like this work?
```
    add_link_options(
        $<$<CONFIG:Debug>:/DEBUG>
        $<$<CONFIG:RelWithDebInfo>:/DEBUG>
    )
```

Why do we add the flag to `CMAKE_EXE` but not to `CMAKE_SHARED` flags? 

Sorry, I think I forgot, why it's a bad approach to specify this flag as global `CMAKE_EXE_LINKER_FLAGS` if build type specified is not Release?

ablekhman replied:
Instead of setting directory properties with `add_link_options` call, I prefill standard CMake variables with additional flags we want to use. This way our additional flags are visible to the user and can be edited before generating make files.

ivankochin replied:
Ok, I see, thanks for reminding me!

BTW wouldn't it be simpler to set CMAKE_CXX_FLAG if not Release instead of setting separate variables for all the build types?

```
if (!BUILD_TYPE Release)
    set(CMAKE_CXX_FLAG -ggdb)
```
vs

```
set (CMAKE_CXX_DEBUG_FLAG -ggdb)
set (CMAKE_CXX_RELWITHDEBINFO_FLAG -ggdb)
set (CMAKE_CXX_TSAN_FLAG -ggdb)
```

ivankochin commented on cmake/configs/base.cmake:
```diff
@@ -438,13 +457,15 @@ if(WT_WIN)
 if(ENABLE_ANTITHESIS)
-    add_compile_options(-fsanitize-coverage=trace-pc-guard)
+    foreach(lang C CXX)
+        add_cmake_flag(CMAKE_${lang}_FLAGS -fsanitize-coverage=trace-pc-guard)
```
Why not `add_cmake_compiler_flags`?

ablekhman replied:
Because this function also used with other flags variables: linker flags, for example.

ivankochin replied:
Discussed IRL, that's not about linker flags but for simplisity

ivankochin commented on CMakeLists.txt:
```diff
@@ -87,26 +87,28 @@ set(CMAKE_CXX_EXTENSIONS OFF)
-    elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" OR "${CMAKE_C_COMPILER_ID}" STREQUAL "Clang")
-        add_compile_options(-fcolor-diagnostics)
+    if(GNU_C_COMPILER OR GNU_CXX_COMPILER)
+        add_cmake_flag(CMAKE_C_FLAGS "-fdiagnostics-color=always")
+        add_cmake_flag(CMAKE_CXX_FLAGS "-fdiagnostics-color=always")
```
Why not `add_cmake_compiler_flags`?

ablekhman replied:
Because it will be an overkill here. Setting `CMAKE_C[XX]_FLAGS` is enough to make it common to all build variation: `*_DEBUG`, `*_RELEASE`, etc.

ivankochin replied:
I just think that it could be useful to have one function that would take care of setting one flag to avoid code duplication in such cases, something like:

```
add_cmake_compiler_flags("-fdiagnostics-color=always") # set it for all build_types and both C and CXX
```

ivankochin commented on cmake/helpers.cmake:
```diff
@@ -589,23 +589,88 @@ function(parse_filelist_source filelist output_var)
+    if(NOT COMPILER_FLAGS_FLAGS)
+        message(FATAL_ERROR "add_cmake_compiler_flags: FLAGS argument is required")
+    endif()
+    if(NOT COMPILER_FLAGS_LANGUAGES)
```
Can we set default for LANGUAGES to C,CXX?

ablekhman replied:
Probably we could, but I'd like to keep the function simple for the sake of the future maintainers. Also, this way the call is explicit about languages.

ivankochin commented on cmake/helpers.cmake:
```diff
@@ -589,23 +589,88 @@ function(parse_filelist_source filelist output_var)
+        message(FATAL_ERROR "add_cmake_compiler_flags: LANGUAGES argument is required")
+    endif()
+    if(NOT COMPILER_FLAGS_BUILD_TYPES)
+        message(FATAL_ERROR "add_cmake_compiler_flags: BUILD_TYPES argument is required")
```
Can we set default BUILD_TYPES to all types that we have?

ablekhman replied:
As above, I rather keep the function simpler, without predefined defaults.

ivankochin commented on cmake/configs/modes.cmake:
```diff
@@ -205,23 +205,24 @@ endif()
+        REMOVE ${opt_flags}
+        ADD ${CC_OPTIMIZE_LEVEL})
+    replace_compile_options(CMAKE_${lang}_FLAGS_RELWITHDEBINFO
+        REMOVE ${opt_flags} ${debug_flags}
```
Should we remove ${debug_flags} from DEBUG build type?

I think we should also do all these replacements for sanitizer build types.

Can we replace debug in another place together with setting our custom debug flags for consistency? Something like:

```
    replace_compile_options(CMAKE_${lang}_FLAGS_RELWITHDEBINFO
        REMOVE ${debug_flags}
        ADD ${WT_CUSTOM_DEBUG_FLAGS})
```

ablekhman replied:
I'm not sure why only these build types were selected. I can see that all sanitiser builds do not have optimisation flags to begin with. Most likely because sanitiser build modes created from scratch without being pre-initialised with default CMake settings.

ivankochin replied:
Thanks for clarifying!

What about Debug build type? and replacing it in a different place?

ivankochin replied:
Still actual questions here

ablekhman replied:
I tried to move the code around, but without major refactor every other solution looks even worse:

- Moving this code to helpers.cmake is incorrect
- Moving to modes.cmake is too early, because build options are defined in base.cmake. These build options will eventually control what goes into build and what doesn't.

ablekhman replied:
About removing the flags for Debug type. I'll need to figure out why only Release* types were selected to begin with. I try to keep the functionality as it was before.

ablekhman committed:
WT-15495 Expose extra debug compilation flags to user

- Extra compilation flags now added to the standard
    `CMAKE_<LANG>_FLAGS_<CONFIG>` variables instead
    of being silently applied to directories
- Flags are visible to the user during config phase
- User can edit the flags, if needed

ivankochin commented on cmake/configs/base.cmake:
```diff
@@ -400,23 +399,49 @@ endif()
-        if("${CMAKE_C_COMPILER_ID}" STREQUAL "Clang")
-            # Clang requires one additional flag to output macro debug information.
-            add_compile_options(-fdebug-macro)
+    set(BUILD_TYPES_WITH_DEBUG_INFO ${BUILD_MODES})
```
np for consistency: 
```suggestion
    set(BUILD_MODES_WITH_DEBUG_INFO ${BUILD_MODES})
```
or should we rename BUILD_MODES to BUILD_TYPES?

ivankochin commented on cmake/configs/base.cmake:
```diff
@@ -400,23 +399,49 @@ endif()
+                    string(TOUPPER "${build_type}" build_type_upper)
+                    add_cmake_flag(CMAKE_${binary_type}_LINKER_FLAGS_${build_type_upper} "/DEBUG")
+                endforeach()
+            endforeach()
```
Can we move it to some function to improve code readability? e.g. `add_cmake_linker_flags` to avoid having all these for loops in base.cmake?

ivankochin reviewed (commented):
Thank you for doing this PR! I've left another bunch of comments, however, I understand that it's better to make changes iteratively so none of them are mandatory to be fixed before merging

ablekhman committed:
WT-15495 Expose extra debug compilation flags to user

- Extra compilation flags now added to the standard
    `CMAKE_<LANG>_FLAGS_<CONFIG>` variables instead
    of being silently applied to directories
- Flags are visible to the user during config phase
- User can edit the flags, if needed

ivankochin commented on cmake/configs/base.cmake:
```diff
@@ -399,25 +398,44 @@ if (HAVE_DIAGNOSTIC)
-        set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /DEBUG")
-    else()
+# Only set initial debug flags once to preserve user customizations.
+if(ENABLE_DEBUG_INFO AND NOT WT_DEBUG_FLAGS_INITIALIZED)
```
Why do we need protection by `WT_DEBUG_FLAGS_INITIALIZED`? 

Are there cases when we call this place twice? 

Can we check `DEBUG_INFO_FLAGS` to avoid introducing an extra variable?

ablekhman replied:
I use `WT_DEBUG_FLAGS_INITIALIZED` flag to make sure that subsequent configure runs won't overwrite user's changes. For example:

1. User runs cmake on clean project; standard cmake var will contain our debug flags.
2. User checks the flags and decides to change them (add more flags, replace, etc).
3. User runs cmake configure step again with his changes. If we don't protect the user's changes at this run, then our code will add debug flags again.

ivankochin reviewed (commented):
Another thing is that since we don't fix all the places by this PR, can we create somewhere a short document of what is the current state of CMake, what problems do we have there, and what are next steps to fix this to make it simpler for future generations to catch up on the decisions that were made.

Doesn't have to be a part of this PR though.
